source/world/gamemode/CreativeMode.cpp

```diff
@@ -9,31 +9,20 @@
 #include "CreativeMode.hpp"
 #include "client/app/Minecraft.hpp"
 
-CreativeMode::CreativeMode(Minecraft* pMC) : GameMode(pMC)
+CreativeMode::CreativeMode(Minecraft* pMC, Level& level) : GameMode(pMC, level)
```

Comment:
I don't understand why you are reforming the game mode system. I have an idea that doesn't involve as much change: simply create a new `GameMode` object (`CreativeMode` or `SurvivalMode`) everytime a level is loaded or a `StartGamePacket` is received.

The way they're implemented in `master` makes me think that these actually belong in `client/player/` though... kind of an oversight on my part

Reply:
Because this is the way that Mojang did it. As far as I am aware, we're still creating one GameMode object within an instance of Minecraft.